File: packages/plplot/examples/c/x18c.c
/* $Id: x18c.c 3186 2006-02-15 18:17:33Z slbrow $

	3-d line and point plot demo.  Adapted from x08c.c.
*/

#include "plcdemos.h"

static int opt[] = { 1, 0, 1, 0 };
static PLFLT alt[] = {20.0, 35.0, 50.0, 65.0};
static PLFLT az[] = {30.0, 40.0, 50.0, 60.0};

void test_poly(int k);

/*--------------------------------------------------------------------------*\
 * main
 *
 * Does a series of 3-d plots for a given data set, with different
 * viewing options in each plot.
\*--------------------------------------------------------------------------*/

#define NPTS 1000

int
main(int argc, char *argv[])
{
    int i, j, k;
    PLFLT *x, *y, *z;
    PLFLT r;
    char title[80];

/* Parse and process command line arguments */

    (void) plparseopts(&argc, argv, PL_PARSE_FULL);

/* Initialize plplot */

    plinit();

    for( k=0; k < 4; k++ )
	test_poly(k);

    x = (PLFLT *) malloc(NPTS * sizeof(PLFLT));
    y = (PLFLT *) malloc(NPTS * sizeof(PLFLT));
    z = (PLFLT *) malloc(NPTS * sizeof(PLFLT));

/* From the mind of a sick and twisted physicist... */

    for (i = 0; i < NPTS; i++) {
	z[i] = -1. + 2. * i / NPTS;

/* Pick one ... */

/*	r    = 1. - ( (PLFLT) i / (PLFLT) NPTS ); */
	r    = z[i];

	x[i] = r * cos( 2. * PI * 6. * i / NPTS );
	y[i] = r * sin( 2. * PI * 6. * i / NPTS );
    }

    for (k = 0; k < 4; k++) {
	pladv(0);
	plvpor(0.0, 1.0, 0.0, 0.9);
	plwind(-1.0, 1.0, -0.9, 1.1);
	plcol0(1);
	plw3d(1.0, 1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, alt[k], az[k]);
	plbox3("bnstu", "x axis", 0.0, 0,
	       "bnstu", "y axis", 0.0, 0,
	       "bcdmnstuv", "z axis", 0.0, 0);

	plcol0(2);

	if (opt[k])
	    plline3( NPTS, x, y, z );
	else
	    plpoin3( NPTS, x, y, z, 1 );

	plcol0(3);
	sprintf(title, "#frPLplot Example 18 - Alt=%.0f, Az=%.0f",
		alt[k], az[k]);
	plmtex("t", 1.0, 0.5, 0.5, title);
    }

    /* Clean up */
    free((void *) x);
    free((void *) y);
    free((void *) z);

    plend();

    exit(0);
}

void test_poly(int k)
{
    PLFLT *x, *y, *z;
    int i, j;
    PLFLT pi, two_pi;
    PLINT draw[][4] = { { 1, 1, 1, 1 },
		      { 1, 0, 1, 0 },
		      { 0, 1, 0, 1 },
		      { 1, 1, 0, 0 } };

    pi = PI, two_pi = 2. * pi;

    x = (PLFLT *) malloc(5 * sizeof(PLFLT));
    y = (PLFLT *) malloc(5 * sizeof(PLFLT));
    z = (PLFLT *) malloc(5 * sizeof(PLFLT));

    pladv(0);
    plvpor(0.0, 1.0, 0.0, 0.9);
    plwind(-1.0, 1.0, -0.9, 1.1);
    plcol0(1);
    plw3d(1.0, 1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, alt[k], az[k]);
    plbox3("bnstu", "x axis", 0.0, 0,
	   "bnstu", "y axis", 0.0, 0,
	   "bcdmnstuv", "z axis", 0.0, 0);

    plcol0(2);

#define THETA(a) (two_pi * (a) /20.)
#define PHI(a)    (pi * (a) / 20.1)

/*
  x = r sin(phi) cos(theta)
  y = r sin(phi) sin(theta)
  z = r cos(phi)
  r = 1 :=)
  */

    for( i=0; i < 20; i++ ) {
	for( j=0; j < 20; j++ ) {
	    x[0] = sin( PHI(j) ) * cos( THETA(i) );
	    y[0] = sin( PHI(j) ) * sin( THETA(i) );
	    z[0] = cos( PHI(j) );
	    
	    x[1] = sin( PHI(j+1) ) * cos( THETA(i) );
	    y[1] = sin( PHI(j+1) ) * sin( THETA(i) );
	    z[1] = cos( PHI(j+1) );
	    
	    x[2] = sin( PHI(j+1) ) * cos( THETA(i+1) );
	    y[2] = sin( PHI(j+1) ) * sin( THETA(i+1) );
	    z[2] = cos( PHI(j+1) );
	    
	    x[3] = sin( PHI(j) ) * cos( THETA(i+1) );
	    y[3] = sin( PHI(j) ) * sin( THETA(i+1) );
	    z[3] = cos( PHI(j) );
	    
	    x[4] = sin( PHI(j) ) * cos( THETA(i) );
	    y[4] = sin( PHI(j) ) * sin( THETA(i) );
	    z[4] = cos( PHI(j) );

	    plpoly3( 5, x, y, z, draw[k], 1 );
	}
    }

    plcol0(3);
    plmtex("t", 1.0, 0.5, 0.5, "unit radius sphere" );

    /* Clean up */
    free((void *) x);
    free((void *) y);
    free((void *) z);

}

File: packages/plplot/bindings/octave/demos/x18c.m
## Copyright (C) 1998, 1999, 2000  Joao Cardoso
## Copyright (C) 2004  Rafael Laboissiere
## 
## This program is free software; you can redistribute it and/or modify it
## under the terms of the GNU General Public License as published by the
## Free Software Foundation; either version 2 of the License, or (at your
## option) any later version.
## 
## This program is distributed in the hope that it will be useful, but
## WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
## General Public License for more details.
##
## This file is part of plplot_octave.
## It is based on the corresponding demo function of PLplot.

1;

global alt = [20.0, 35.0, 50.0, 65.0];
global az = [30.0, 40.0, 50.0, 60.0];

##  Does a series of 3-d plots for a given data set, with different
##  viewing options in each plot.

function ix18c

  if (!exist("plinit"))
    plplot_stub
  endif

  if(isempty(sprintf("%s",plgdev')))
    plsdev("xwin");
  endif

  global alt;
  global az;
  global opt = [ 1, 0, 1, 0 ];
  NPTS = 1000;

  ## Parse and process command line arguments */

  ##    (void) plParseOpts(&argc, argv, PL_PARSE_FULL);

  ## Initialize plplot */
  plinit();

  for k=0:3
    test_poly(k);
  endfor

  ## From the mind of a sick and twisted physicist... */

  for i = 0:NPTS-1
    z(i+1) = -1. + 2. * i / NPTS;

    ## Pick one ... */
				#	r    = 1. - ( (float) i / (float) NPTS ); */
    r    = z(i+1);

    x(i+1) = r * cos( 2. * pi * 6. * i / NPTS );
    y(i+1) = r * sin( 2. * pi * 6. * i / NPTS );
  endfor

  for k = 0:3
    pladv(0);
    plvpor(0.0, 1.0, 0.0, 0.9);
    plwind(-1.0, 1.0, -0.9, 1.1);
    plcol(1);
    plw3d(1.0, 1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, alt(k+1), az(k+1));
    plbox3("bnstu", "x axis", 0.0, 0,
	   "bnstu", "y axis", 0.0, 0,
	   "bcdmnstuv", "z axis", 0.0, 0);

    plcol(2);

    if (opt(k+1))
      plline3( x', y', z' );
    else
      plpoin3( x', y', z', 1 );
    endif

    plcol(3);
    title=sprintf("#frPLplot Example 18 - Alt=%.0f, Az=%.0f",
		  alt(k+1), az(k+1));
    plmtex("t", 1.0, 0.5, 0.5, title);
  endfor

  plend1();
endfunction

function y = THETA(a)
 y = 2 * pi * (a) /20.;
endfunction

function y= PHI(a)
  y = pi * (a) / 20.1;
endfunction

function test_poly(k)

  global alt;
  global az;
  
  draw = [ 1, 1, 1, 1;
	  1, 0, 1, 0;
	  0, 1, 0, 1;
	  1, 1, 0, 0];

  two_pi = 2. * pi;

  pladv(0);
  plvpor(0.0, 1.0, 0.0, 0.9);
  plwind(-1.0, 1.0, -0.9, 1.1);
  plcol(1);
  plw3d(1.0, 1.0, 1.0, -1.0, 1.0, -1.0, 1.0, -1.0, 1.0, alt(k+1), az(k+1));
  plbox3("bnstu", "x axis", 0.0, 0,
	 "bnstu", "y axis", 0.0, 0,
	 "bcdmnstuv", "z axis", 0.0, 0);

  plcol(2);


  for i=0:19
    for j=0:19
      
      pj=pi*j/20.1;	pj1=pi*(j+1)/20.1;
      ti=2*pi*i/20; ti1=2*pi*(i+1)/20;
      
      x(1) = sin( pj ) * cos( ti );
      y(1) = sin( pj ) * sin( ti );
      z(1) = cos( pj );
      
      x(2) = sin( pj1 ) * cos( ti );
      y(2) = sin( pj1 ) * sin( ti );
      z(2) = cos( pj1 );
      
      x(3) = sin( pj1 ) * cos( ti1 );
      y(3) = sin( pj1 ) * sin( ti1 );
      z(3) = cos( pj1 );
      
      x(4) = sin( pj ) * cos( ti1 );
      y(4) = sin( pj ) * sin( ti1 );
      z(4) = cos( pj );
      
      x(5) = sin( pj ) * cos( ti );
      y(5) = sin( pj ) * sin( ti );
      z(5) = cos( pj );

      plpoly3(x', y', z', draw(k+1,:)', -1); ## added an extra argument, with the sign 
    endfor
  endfor

  plcol(3);
  plmtex("t", 1.0, 0.5, 0.5, "unit radius sphere" );
endfunction

ix18c
